btf/btf.go

```diff
@@ -487,6 +531,37 @@ func fixupDatasec(types []Type, sectionSizes map[string]uint32, offsets map[symb
 	return nil
 }
 
+// fixupDatasecLayout populates ds.Vars[].Offset according to var sizes and
+// alignment. Calculate and set ds.Size.
+func fixupDatasecLayout(ds *Datasec) error {
+	var off uint32
+
+	for i, vsi := range ds.Vars {
+		v := vsi.Type.(*Var)
+		size, err := Sizeof(v.Type)
+		if err != nil {
+			return fmt.Errorf("variable %s: getting size: %w", v.Name, err)
+		}
+		align, err := alignof(v.Type)
+		if err != nil {
+			return fmt.Errorf("variable %s: getting alignment: %w", v.Name, err)
+		}
+
+		// Calculate the padding (if any) between the end of the previous member
+		// and the start of the current member. Advance the offset, store it in vsi
+		// and increment by the current member's size.
+		off += padding(off, uint32(align))
+
+		ds.Vars[i].Offset = off
+
+		off += uint32(size)
+	}
+
+	ds.Size = uint32(off)
```

Comment:
Idle curiosity: does the kernel require datasec size also have to be a multiple of 4 or something like that?

Reply:
Likely not, Datasec size can be 5 if it contains a u32 followed by a u8, for example.